### PR TITLE
ARROW-13826: [C++][Gandiva] Implement QUOTE Hive functions on Gandiva

### PR DESCRIPTION
**Pull request by jpedroantunes**
Implement QUOTE Hive functions on Gandiva

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops_test.cc [hunk 1]
```diff
@@ -912,6 +912,33 @@ TEST(TestStringOps, TestReverse) {
...
+
+  out_str = quote_utf8(ctx_ptr, "don't", 5, &out_len);
+  EXPECT_EQ(std::string(out_str, out_len), "\'don\\'t\'");
+  EXPECT_FALSE(ctx.has_error());
+
+  out_str = quote_utf8(ctx_ptr, "", 0, &out_len);
+  EXPECT_EQ(std::string(out_str, out_len), "");
+  EXPECT_FALSE(ctx.has_error());
+
+  out_str = quote_utf8(ctx_ptr, "'", 1, &out_len);
```
Comment (rkavanap) [inline cpp/src/gandiva/precompiled/string_ops_test.cc:937]: this is a good test. Maybe another one like this "''''''''''''''''''''''" with more quotes?

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 2]
```diff
@@ -1762,6 +1762,41 @@ const char* replace_utf8_utf8_utf8(gdv_int64 context, const char* text,
...
+    if (memcmp(in + i, "'", 1) == 0) {
+      out[counter] = '\\';
+      counter++;
+      out[counter] = '\'';
+    } else {
+      out[counter] = in[i];
+    }
+    counter++;
+  }
+  out[counter] = '\'';
```
Comment (rkavanap) [inline cpp/src/gandiva/precompiled/string_ops.cc:1795]: won't this overflow for the case we have all quotes. Say we have 2 quotes and in_len is 2.. we allocate only 4 bytes, but this logic may need 5 bytes as follows '\'\'' which is 5 chars and not 4.

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 3]
```diff
@@ -1762,6 +1762,41 @@ const char* replace_utf8_utf8_utf8(gdv_int64 context, const char* text,
...
+//      DON'T -> 'DON\'T'
+FORCE_INLINE
+const char* quote_utf8(gdv_int64 context, const char* in, gdv_int32 in_len,
+                       gdv_int32* out_len) {
+  if (in_len <= 0) {
+    *out_len = 0;
+    return "";
+  }
+  // try to allocate double size output string (worst case)
+  auto out = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, in_len * 2));
```
Comment (rkavanap) [inline cpp/src/gandiva/precompiled/string_ops.cc:1776]: shouldn't this be in_len * 2 + 1 for the worst case of the loop?

**Review by rkavanap — CHANGES_REQUESTED** [review]

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 2 above)
Comment (rkavanap) [inline cpp/src/gandiva/precompiled/string_ops.cc:1795]: won't this overflow for a quote only string. let us say we have a string with single quote of len 1,, then out is of len 2, but counter could be 3.

File: cpp/src/gandiva/precompiled/string_ops_test.cc (hunk 1 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops_test.cc:937]: Added a test with more quotes

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 3 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc:1776]: Yes, it makes sense, added the +2 for the worst case of the loop

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 2 above)
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc:1795]: Out len should be three as the out string will be 3 quotes, as the input single quote should be quoted

**Review by rkavanap — APPROVED** [review]

**Pull request closed without merging**